dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/cloudera/manager/ClouderaClustersTaskTest.java

```diff
@@ -0,0 +1,271 @@
+    when(context.getArguments()).thenReturn(arguments);
+
+    apiClusterListJson = readString("/cloudera/manager/dto/ApiClusterList.json");
+    apiClusterJson = readString("/cloudera/manager/dto/ApiCluster.json");
+    clusterStatusJson = readString("/cloudera/manager/dto/cluster-status.json");
+  }
+
+  @Test
+  public void clusterNotProvided_success() throws Exception {
+    when(arguments.getCluster()).thenReturn(null);
+
+    try (MockedStatic<EntityUtils> mockedUtils = mockStatic(EntityUtils.class)) {
+      CloseableHttpResponse clustersResponse = mock(CloseableHttpResponse.class);
+      HttpEntity clustersEntity = mock(HttpEntity.class);
+      when(clustersResponse.getEntity()).thenReturn(clustersEntity);
+      when(httpClient.execute(
+              argThat(get -> get != null && get.getURI().toString().endsWith("/clusters"))))
+          .thenReturn(clustersResponse);
+      mockedUtils.when(() -> EntityUtils.toString(clustersEntity)).thenReturn(apiClusterListJson);
+
+      // request for cluster aaa
+      CloseableHttpResponse statusAResponse = mock(CloseableHttpResponse.class);
+      HttpEntity statusAEntity = mock(HttpEntity.class);
+      when(statusAResponse.getEntity()).thenReturn(statusAEntity);
+      when(httpClient.execute(
+              argThat(get -> get != null && get.getURI().toString().endsWith("/aaa/status.json"))))
+          .thenReturn(statusAResponse);
+      mockedUtils
+          .when(() -> EntityUtils.toString(statusAEntity))
+          .thenReturn(clusterStatusJson.replaceAll("" + clusterStatusId, "111"));
+      when(statusAResponse.getStatusLine())
+          .thenReturn(new BasicStatusLine(HttpVersion.HTTP_1_1, 200, "Ok"));
+
+      // request for cluster bbb
+      CloseableHttpResponse statusBResponse = mock(CloseableHttpResponse.class);
+      HttpEntity statusBEntity = mock(HttpEntity.class);
+      when(statusBResponse.getEntity()).thenReturn(statusBEntity);
+      when(httpClient.execute(
+              argThat(get -> get != null && get.getURI().toString().endsWith("/bbb/status.json"))))
+          .thenReturn(statusBResponse);
+      mockedUtils
+          .when(() -> EntityUtils.toString(statusBEntity))
+          .thenReturn(clusterStatusJson.replaceAll("" + clusterStatusId, "222"));
+      when(statusBResponse.getStatusLine())
+          .thenReturn(new BasicStatusLine(HttpVersion.HTTP_1_1, 401, "not found"));
+
+      task.doRun(context, sink, handle);
+
+      verify(writer)
+          .write(
+              (String)
+                  argThat(
+                      content -> {
+                        try {
+                          ApiClusterListDto listDto =
+                              objectMapper.readValue((String) content, ApiClusterListDto.class);
+                          assertNotNull(listDto.getItems());
+                        } catch (JsonProcessingException e) {
+                          throw new RuntimeException(e);
```

Comment:
This test makes some assertions, but it also makes a check without assertions, following the steps:
- Make a conversion from format A to B
- Convert the results back from B to A
- Check that no exceptions were thrown

This kind of check can detect some errors, but doesn't verify or document much about what format B actually looks like. It would be very helpful to have a more typical test:
- Take an input in format A
- Make a one-way conversion
- Compare the result with expected output in format B

Rewriting existing tests is not required, but I'd ask you to add a test of the second type for this logic.

Reply:
Could you please clarify why this test: `doesn't verify or document much about what format B actually looks like` ?

This test asserts:
1. The string in file is valid `json`
2. The `json` can be parsed into structure `ApiClusterListDto` with required fields (only one to be honest).

The order of fields in `json` or formatting with tabs or spaces is not important it's why it's not asserted here.

Reply:
Resolved. Tests are valid, however I agree that readability is not the best one. If any idea for improvement will come, let's consider them in the scope of http://b/377651080